compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -696,22 +697,45 @@ trait Implicits { self: Typer =>
         }
     }
     def location(preposition: String) = if (where.isEmpty) "" else s" $preposition $where"
+    def userDefinedMessage(annot: Annotation, params: List[String], args: List[Type]): Option[String] =
+      for (Trees.Literal(Constant(raw: String)) <- annot.argument(0)) yield {
```

Comment:
Wondering about the magic constant `0`, I guess because the string is the 0th argument for all annotations currently supported? But if you try to make `userDefinedMessage` a good abstraction, *maybe* `annot.argument(0)` should move to the call site (even if, indeed, it gets duplicated).